### ISSUE TITLE
coordinates: ICRSCoordinates should accept coordinate strings with no space

### ISSUE THREAD

**Issue by adrn** [astropy/astropy#1115]
The `ICRSCoordinates` initializer should probably accept:

    ICRSCoordinates("J061800.25+203142.5")
    ICRSCoordinates("061800.25+203142.5", unit=(u.hour,u.degree))

Right now these fail for a number of reasons:
* the preceding J 
* no space in the coordinate string
* the angle parser can't handle angles specified without a delimiter, e.g.: 061800.25

I started working on fixing this, but got lost in the "new" pyparsing code...

**Comment by astrofrog** [comment]
Closed by https://github.com/astropy/astropy/pull/2920

**Comment by astrofrog** [comment]
Sorry, this issue is not completely addressed by #2920 (there was a comment there that said it closed this issue). The strings here are still not parsed, and I'm not sure if we decided we should or not. There was a comment above by @taldcroft that we should support IAU names at least. Removing the 1.0 milestone.

**Comment by bsipocz** [comment]
I've looked at these formats while doing #2920. The first one is easy, one must really need to use the unit as both ``u.deg`` or ``u.hour`` can be valid unit of RA for this format.

Regarding the second, I wasn't sure whether to include them with hard wired fix lenght as HHMMSS.ss+DDMMSS.s and (D)DDMMSS.ss+DDMMSS.s. 

The current example is OK, but it can be quite ambiguous when one says ``"000010.25+000055.3"`` as it can be parsed as a decimal ``u.hour`` and ``u.deg``. However no one would write that many zeros in front of a decimal value... 

btw is there an official IAU format/name list somewhere? I only found examples of supported and avoidable formats, but nothing comprehensive.

**Comment by adrn** [comment]
@bsipocz @astrofrog  [This IAU spec document](http://cdsweb.u-strasbg.fr/Dic/iau-spec.html) states that RA is always specified in HMS, so now I don't think there is a unit ambiguity. In this document, it doesn't look like there is much flexibility and this could just be implemented as a regex... The relevant text from the linked document:

> 
* Equatorial Coordinates shall always be preceded by J if they are for the standard equinox of J2000.0 (i.e., IRCS position or FK5-based, Julian equinox 2000.0 system). They should be preceded by B if they are for the old standard equinox of B1950.0 (i.e., Bessel-Newcomb FK4-based, Besselian equinox 1950.0 system). Galactic coordinates shall be preceded by a G
* Coordinates shall be specified as LLL.ll+BB.bb or LLL.ll-BB.bb for galactic coordinates, and as HHMMSS.ss+DDMMSS.s or HHMMSS.ss-DDMMSS.s for equatorial coordinates (without spaces)
* Coordinates shall be truncated (not rounded), thus defining a unique (small) field on the sky in which the source is located. The truncation should also operate when the right-most digit represents a decimal part. The right-most digit of the field HHMMm should be computed as m=int(SS/6)

I think this regex captures everything outlined in the document:

    (^|\s)([JB](\d{6}|\d{4}|\d{2})\.*\d*[\+\-](\d{6}|\d{4}|\d{2})\.*\d*)|(G\d{3}\.*\d*[\+\-]\d{2}\.*\d*)

e.g., try it on these examples:

```markdown
J061800.25+203142.5
SDSS J061800.25+203142.5
RX J1426.8+6950
PSR J1302-6350
PN G001.2-00.3
QSO B004848-4242.8
QSO B00488-4242
```

**Comment by mhvk** [comment]
Yes, we should still have a parser!  But be sure to also capture odd numbers of digits to get coordinates like `QSO^0048-427`
> If the number of digits is odd and fewer than six, the right-most digit represents a decimal part of hours, degrees or minutes (as, e.g., in the PKS-style HHMM+DDd or in IRAS source designation HHMMm+DDMM) and not tens of minutes or seconds (e.g.. the formats HHMMS or +DDM should be avoided).

**Comment by adrn** [comment]
I'll have to look in to implementing this in a parser -- I have no experience with that and don't understand @mdboom 's original comment!

> I think the best approach will be to create a new parser that imports productions from the existing angle parser. (I don't know if this is even possible with PLY).

**Comment by bsipocz** [comment]
@adrn - If you prefer I can look into this and do PR in the weekend, as I recall in #2920 my major problem was that there wasn't a complete list what needed to be accepted and what shouldn't.

**Comment by adrn** [comment]
@bsipocz aha! I think we just need to modify the code from that PR (#2920) slightly to handle all cases that are described in the document above. If you feel up to it, go for it! Otherwise I can take a crack at it in a few days.

**Comment by bsipocz** [comment]
@adrn - Not before the weekend, but then I probably can do it.

**Comment by adrn** [comment]
Pinging on this! @bsipocz

**Comment by taldcroft** [comment]
I'm good with the general idea of this now.  I just have to laugh/cry when I see the contortions that ensue from this insistence on sexagesimal.  Is it the intent here to fully implement this IAU spec?
 
*Coordinates using an even number of digits (in either RA or Dec), fewer than seven, are expressed in the sexagesimal system. The sequences HHMM.mm or DD.dd where mm and dd are decimal parts of a minute or degree, respectively, should be avoided. If the number of digits is odd and fewer than six, the right-most digit represents a decimal part of hours, degrees or minutes (as, e.g., in the PKS-style HHMM+DDd or in IRAS source designation HHMMm+DDMM) and not tens of minutes or seconds (e.g.. the formats HHMMS or +DDM should be avoided). If the number of digits is more than six, the digits in excess of six are decimal parts of seconds of time for RA or of angle for Dec ; explicit use of the decimal points is encouraged (e.g., HHMMSS.ss or DDMMSS.s).*

**Comment by mhvk** [comment]
The contortions mostly come about because both radio and X-ray astronomers had so many coordinates of poor quality... Anyway, I do think if we implement the specification, we might as well implement it fully. (Though we cannot avoid that, surely, people regularly will continue to assume coordinates are rounded rather than truncated...)

**Comment by adrn** [comment]
Reviving this thread! @bsipocz are you still interested in trying this?

**Comment by bsipocz** [comment]
@adrn - Indeed, this slipped through multiple times. I try to get back to it this week.

(and I wish github would finally introduce a personal todo to help to avoid cases like this...)

**Comment by adrn** [comment]
👋 O hai @bsipocz

**Comment by bsipocz** [comment]
ouch
